libs/mpi/src/mpi_future.cpp

```diff
@@ -0,0 +1,434 @@
+                keep_trying = false;
+
+                if (mpi_debug.is_enabled())
+                {
+                    auto poll_deb = mpi_debug.make_timer(1,
+                        debug::str<>("Poll - <ERR>"), detail::get_mpi_info());
+
+                    // clang-format off
+                    mpi_debug.error(poll_deb,
+                        debug::str<>("Poll <ERR>"),
+                        "MPI_ERROR", detail::error_message(status.MPI_ERROR),
+                        "status", debug::dec<>(status.MPI_ERROR),
+                        "index", debug::dec<>(index),
+                        "flag", debug::dec<>(flag));
+                    // clang-format on
+                }
+            }
+        }
+
+        // if there are more than 25% NULL request handles in our lists, compact them
+        if (detail::get_active_futures().size() > 0)
+        {
+            std::size_t nulls =
+                std::count(detail::get_active_requests().begin(),
+                    detail::get_active_requests().end(), MPI_REQUEST_NULL);
+
+            if (nulls > detail::get_active_requests().size() / 4)
+            {
+                // compact away any requests that have been replaced by
+                // MPI_REQUEST_NULL
+                auto end1 = std::remove(detail::get_active_requests().begin(),
+                    detail::get_active_requests().end(), MPI_REQUEST_NULL);
+                detail::get_active_requests().resize(
+                    std::distance(detail::get_active_requests().begin(), end1));
+
+                // compact away any null pointers in futures list
+                auto end2 = std::remove(detail::get_active_futures().begin(),
+                    detail::get_active_futures().end(), nullptr);
+                detail::get_active_futures().resize(
+                    std::distance(detail::get_active_futures().begin(), end2));
+
+                if (detail::get_active_requests().size() !=
+                    detail::get_active_futures().size())
+                {
+                    HPX_THROW_EXCEPTION(invalid_status, "hpx::mpi::poll",
+                        "Fatal Error: Mismatch in vectors");
+                }
+
+                if (mpi_debug.is_enabled())
+                {
+                    mpi_debug.debug(debug::str<>("MPI_REQUEST_NULL"),
+                        detail::get_mpi_info(), "list size",
+                        debug::dec<3>(detail::get_active_futures().size()),
+                        "nulls ", debug::dec<>(nulls));
+                }
+            }
+        }
+    }
+
+    namespace detail {
+
+        void register_polling(hpx::threads::thread_pool_base& pool)
+        {
+            auto* sched = pool.get_scheduler();
+
+            mpi_debug.debug(debug::str<>("Setting mode"),
+                detail::get_mpi_info(), "enable_user_polling");
+
+            // always set polling function before enabling polling
+            sched->set_user_polling_function(&hpx::mpi::poll);
```

Comment:
It would be nice if this weren't tied to the scheduler and could be done with just plain HPX threads. If another executor will also silently set the user polling function, it should at least support setting multiple such functions. That's for outside this PR though.

Reply:
Yes, I agree. Let's leave this for another PR though. I'll create a ticket for this.